modules/openxr/extensions/openxr_composition_layer_extension.cpp

```diff
@@ -82,42 +82,96 @@ void OpenXRCompositionLayerExtension::on_session_created(const XrSession p_sessi
 
 void OpenXRCompositionLayerExtension::on_session_destroyed() {
 	OpenXRAPI::get_singleton()->unregister_composition_layer_provider(this);
-
-#ifdef ANDROID_ENABLED
-	free_queued_android_surface_swapchains();
-#endif
 }
 
 void OpenXRCompositionLayerExtension::on_pre_render() {
-#ifdef ANDROID_ENABLED
-	free_queued_android_surface_swapchains();
-#endif
-
-	for (OpenXRViewportCompositionLayerProvider *composition_layer : composition_layers) {
+	for (CompositionLayer *composition_layer : registered_composition_layers) {
 		composition_layer->on_pre_render();
 	}
 }
 
 int OpenXRCompositionLayerExtension::get_composition_layer_count() {
-	return composition_layers.size();
+	return registered_composition_layers.size();
 }
 
 XrCompositionLayerBaseHeader *OpenXRCompositionLayerExtension::get_composition_layer(int p_index) {
-	ERR_FAIL_INDEX_V(p_index, composition_layers.size(), nullptr);
-	return composition_layers[p_index]->get_composition_layer();
+	ERR_FAIL_UNSIGNED_INDEX_V((unsigned int)p_index, registered_composition_layers.size(), nullptr);
+	return registered_composition_layers[p_index]->get_composition_layer();
 }
 
 int OpenXRCompositionLayerExtension::get_composition_layer_order(int p_index) {
-	ERR_FAIL_INDEX_V(p_index, composition_layers.size(), 1);
-	return composition_layers[p_index]->get_sort_order();
+	ERR_FAIL_UNSIGNED_INDEX_V((unsigned int)p_index, registered_composition_layers.size(), 1);
+	return registered_composition_layers[p_index]->sort_order;
+}
+
+RID OpenXRCompositionLayerExtension::composition_layer_create(XrCompositionLayerBaseHeader *p_openxr_layer) {
+	RID rid = composition_layer_owner.make_rid();
+	CompositionLayer *layer = composition_layer_owner.get_or_null(rid);
+
+	switch (p_openxr_layer->type) {
+		case XR_TYPE_COMPOSITION_LAYER_QUAD: {
```

Comment:
We need something better for this, if vendors introduce vendor extensions for composition layers, we can't use this system for that. There is also a risk if any composition layer needs things in their next pointers and that data doesn't get copied along.

This is why for spatial entities I introduced `OpenXRStructureBase` as a refcounted wrapper class for these types of structures.

We simply need to define a `OpenXRCompositionLayerBaseHeader` abstract class that subclasses `OpenXRStructureBase` and then create subclasses for each composition layer type.

It's a much cleaner approach, allows for using additional wrapper classes for anything in the next chain, and allows us to support vendor composition layers down the line.